compiler/rustc_borrowck/src/diagnostics/mutability_errors.rs

```diff
@@ -953,6 +955,41 @@ impl<'a, 'tcx> MirBorrowckCtxt<'a, 'tcx> {
         }
     }
 
+    fn suggest_using_iter_mut(&self, err: &mut DiagnosticBuilder<'_, ErrorGuaranteed>) {
+        let source = self.body.source;
+        let hir = self.infcx.tcx.hir();
+        if let InstanceDef::Item(def_id) = source.instance
+            && let Some(Node::Expr( hir::Expr {hir_id, kind, ..})) = hir.get_if_local(def_id)
+            && let ExprKind::Closure(closure) = kind
+            && closure.movability == None
+            && let Some(Node::Expr(expr)) = hir.find_parent(*hir_id) {
+                let mut cur_expr = expr;
+                loop {
+                    if let ExprKind::MethodCall(path_segment, recv, _, _) = cur_expr.kind {
+                        if path_segment.ident.name == sym::iter {
+                            let def_id = hir.enclosing_body_owner(recv.hir_id);
+                            let tables = self.infcx.tcx.typeck(def_id);
+                            let self_ty = tables.node_type_opt(recv.hir_id);
+                            if let Some(_ty) = self_ty {
+                                // FIXME: check `_ty` has `iter_mut` method ?
```

Comment:
This shouldn't be too tricky to do - [here's an example of Esteban doing something similar in a recent PR of his](https://github.com/rust-lang/rust/pull/115147/files#diff-23edf23f7466b97eb3154b82f7b324983287c327937fd66e854b3df6759926ceR817) (from my recollection of reviewing it, didn't spend too much time confirming that it is what you need, but hopefully a starting point).

Reply:
nice!
thanks for the tip, updated the code.